add TemporalAttribute and related classes

njr-11 opened the pull request:
Add TemporalAttribute and related static metamodel classes

Fixes #1022

gavinking commented on api/src/main/java/jakarta/data/metamodel/TemporalExpression.java:
```diff
@@ -0,0 +1,75 @@
+        extends ComparableExpression<T, V> {
+
+    // TODO How much does it matter to be able to supply the entity class?
+    // For now, I included it in one place (here) and not in the other (.of)
```
I think in the below cases the unbound type variable is completely fine, and we don't need to make people pass in a class as a witness.

If only Java had a bottom type!

gavinking replied:
P.S. It might seem I'm being inconsistent here because I insisted on passing the class in other places. But in the other places the type variable represents a real thing (that we now actually reify). Here it doesn't. If we could declare this operation as:

```java
static <T> TemporalExpression<Nothing, LocalDate> currentDate();
```

with `Nothing` being the bottom type, then we would do that. Unfortunately Java's type system is very incomplete.

Also, in the actual usage patterns for these operations, the type argument can always be easily and correctly inferred. We're never going to be assigning `currentDate()` to a `var`.

gavinking replied:
Ohh, wait wait wait, this is contravariant not covariant. 

So a bit item on my todo list was to consider how all this shit is impacted by inheritance. That is, an `Attribute<Book>` is also an `Attribute<AcademicBook>`, and so some operations are currently declared with signatures that are too restrictive.

I have to think a bit hard about that, but one direction this could go in is that, e.g. `greaterThan` might get declared like this:

```java
default Restriction<T> greaterThan(Expression<? super T,V> value)
```

That is, we could allow an attribute value to be compared against a the value of an attribute declared by a supertype. [Unfortunately, there's an asymmetry here which is inelegant.]

And then we could declare `currentDate()` like this:

```java
static TemporalExpression<Object, LocalDate> currentDate();
```

Letting you write `_Book.publicationDate.greaterThan( currentDate());`.

gavinking replied:
>  and so some operations are currently declared with signatures that are too restrictive.

To clarify: I'm talking here about all the "commented-out" operations.

gavinking commented on api/src/main/java/jakarta/data/metamodel/expression/CurrentDate.java:
```diff
@@ -0,0 +1,29 @@
+
+public interface CurrentDate<T> extends TemporalExpression<T,LocalDate> {
+
+    static <T> CurrentDate<T> of() {
```
I would call it `now()` since that what it's called by `LocalDate` and friends.

njr-11 commented:
Rebased onto latest and resolved a soft merge conflict with the removal of `.attribute()`

gavinking commented on api/src/main/java/jakarta/data/metamodel/TemporalExpression.java:
```diff
@@ -0,0 +1,52 @@
+public interface TemporalExpression<T, V extends Temporal & Comparable<? extends Temporal>>
+        extends ComparableExpression<T, V> {
+
+    static <T> TemporalExpression<T, LocalDate> currentDate() {
+        return CurrentDate.now();
+    }
+
+    static <T> TemporalExpression<T, LocalDateTime> currentDateTime() {
+        return CurrentDateTime.now();
+    }
+
+    static <T> TemporalExpression<T, LocalTime> currentTime() {
+        return CurrentTime.now();
+    }
```
```suggestion
    static TemporalExpression<Object, LocalDate> localDate() {
        return CurrentDate.now();
    }

    static TemporalExpression<Object, LocalDateTime> localDateTime() {
        return CurrentDateTime.now();
    }

    static TemporalExpression<Object, LocalTime> localTime() {
        return CurrentTime.now();
    }
```

njr-11 commented:
I merged Gavin's suggestion/review comment to rename the currentDate methods to localDate, after which I rebased onto latest and addressed merge conflicts. This PR is ready for review again.

gavinking reviewed (approved):
LGTM
